### PR TITLE
Layer-wise LR Decay (LLRD) — stabilized deep training across CFD benchmarks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

**Layer-wise Learning Rate Decay (LLRD) to stabilize training depth utilization across CFD benchmarks.**

LLRD was originally introduced for BERT fine-tuning (Sun et al., 2019) and has since become standard practice in vision transformers. The key insight: deeper/earlier layers learn more generic representations that should be updated more slowly, while later/output layers learn task-specific features that need faster adaptation.

The current Transolver uses a uniform learning rate across all layers. With LLRD, each layer gets a decayed version of the base LR:

```
lr_layer_k = base_lr * decay_factor^(num_layers - k - 1)
```

For a 3-layer model with decay=0.75 and base_lr=1.25e-4:
- Layer 0 (input): lr = 1.25e-4 × 0.75² = 7.03e-5
- Layer 1 (middle): lr = 1.25e-4 × 0.75¹ = 9.375e-5  
- Layer 2 (output): lr = 1.25e-4 × 0.75⁰ = 1.25e-4

This is especially relevant because our current configs already use cosine annealing — LLRD is orthogonal to the LR schedule and can be layered on top.

## Implementation

Create per-layer parameter groups with decayed learning rates:

```python
def get_llrd_param_groups(model, base_lr, decay=0.75):
    """
    Build optimizer parameter groups with layer-wise LR decay.
    Layers closer to the output get higher LR; input layers get lower LR.
    """
    # Identify layer indices — adjust attribute names to match actual model structure
    layer_params = []
    num_layers = len(model.blocks)  # or model.layers, model.encoder, etc.
    
    for i, layer in enumerate(model.blocks):
        layer_lr = base_lr * (decay ** (num_layers - i - 1))
        layer_params.append({
            'params': list(layer.parameters()),
            'lr': layer_lr,
            'name': f'layer_{i}'
        })
    
    # Non-layer params (embedding, head, norm) get full base_lr
    layer_param_ids = set(id(p) for group in layer_params for p in group['params'])
    other_params = [p for p in model.parameters() if id(p) not in layer_param_ids]
    layer_params.append({'params': other_params, 'lr': base_lr, 'name': 'head'})
    
    return layer_params

# Usage: replace model.parameters() with get_llrd_param_groups(model, args.lr, decay=0.75)
param_groups = get_llrd_param_groups(model, args.lr, decay=0.75)
optimizer = Lion(param_groups, lr=args.lr, weight_decay=args.weight_decay)
# or AdamW(param_groups, ...) for airfrans/drivaerml
```

Use **decay=0.75** as the starting point. This is the standard value from the ViT fine-tuning literature.

## Runs (3 datasets)

Use `--wandb_group llrd-cross-dataset` on all runs.

### Run 1 — TandemFoil

```bash
python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 1.25e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --ema-decay 0.999 \
  --wandb_group llrd-cross-dataset
```

### Run 2 — AirfRANS

```bash
python train.py \
  --dataset airfrans \
  --airfrans-task full \
  --optimizer adamw \
  --lr 6e-4 \
  --cosine-t-max 50 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --wandb_group llrd-cross-dataset
```

Note: With only 2 layers, LLRD gives: output layer at 6e-4, input layer at 4.5e-4. Still meaningful.

### Run 3 — DrivAerML

```bash
SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group llrd-cross-dataset
```

DrivAerML has 4 layers, so decay=0.75 gives a meaningful spread:
- Layer 0: 5e-4 × 0.75³ = 2.11e-4
- Layer 1: 5e-4 × 0.75² = 2.81e-4
- Layer 2: 5e-4 × 0.75¹ = 3.75e-4
- Layer 3: 5e-4 × 0.75⁰ = 5.0e-4

## Current Baselines

| Dataset | Metric | Current Best | Best PR |
|---------|--------|-------------|---------|
| TandemFoil | val_primary/surface_pressure_mae | 22.537 | #2924 |
| AirfRANS | val_primary/surface_mse | 0.000482 | #2951 |
| DrivAerML | val_primary/surface_rel_l2_pct | 3.997% | #2898 |

## Expected outcome

LLRD stabilizes training by preventing destructive updates to early-layer representations. With the current deep configs (3-4 layers), the gradient signal may be pulling the input layers too aggressively, which LLRD corrects. DrivAerML (4 layers) is most likely to benefit.

## Reporting

Please report:
1. W&B run IDs for all 3 datasets
2. Best val_primary metric for each dataset
3. Loss curves — specifically whether training is smoother with LLRD vs uniform LR
4. If results are mixed: note which dataset benefited and which did not

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
